File: 2_Regression/Regression_andy/Section 8 - Decision Tree Regression/decision_tree_regression.py
# Regression Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

# Fitting the Decision Tree Regression Model to the dataset
# Create your regressor here
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state=0)
regressor.fit(X,y)

# Predicting a new result
y_pred = regressor.predict(6.5) # we get a lower bound answer, this is because..
#..This Decision Tree is a non-continuous & non-linear model. The decision tree..
#..'s leaves have averaged the y-values. But in each leaf we only have a signle..
#..datapoint. So any output will have a coarse prediction assigned. See plot.

# Visualising the Decisiion Tree Regression results (for higher resolution and smoother curve)
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, regressor.predict(X_grid), color = 'blue')
plt.title('Truth or Bluff (Decisiion Tree Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

File: 2_Regression/Regression_andy/Section 8 - Decision Tree Regression/decision_tree_regression.R
# Regression Template

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# # install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 2/3)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting the Regression Model to the dataset
# Create your regressor here
#install.packages("rpart")
library("rpart")

regressor = rpart(formula = dataset$Salary ~ .,
                  data = dataset,
                  control = rpart.control(minsplit = 1) # default is 20,.. 
                  #..but we only have 10 observations, so it will yeild a..
                  #..split that averages a value over all observations if..
                  #..minsplit wasn't set to 1 (allows splits to contain 1 obs)
)

# Predicting a new result
y_pred = predict(regressor, data.frame(Level = 6.5))

# Visualising the Regression Model results
# install.packages('ggplot2')
library(ggplot2)

x_grid = seq(from=min(dataset$Level), to=max(dataset$Level), by=0.01)

ggplot()+
  geom_point(aes(x=dataset$Level,y=dataset$Salary), colour="red")+
  geom_line(aes(x=x_grid,y=predict(regressor, newdata = data.frame(Level = x_grid))), colour="blue")+
  ggtitle("Decision Tree Regression")+
  xlab("Level")+
  ylab("Salary")



